### PR TITLE
Fix as conversion in memfs.rs

### PR DESCRIPTION
**Pull request by pzheng1025**

### CODE REVIEW

File: src/fuse/conversion.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+macro_rules! impl_cast {
+    ($From: ty, $To: ty) => {
+        impl Cast<$From> for $To {
+            #[inline]
+            fn cast(value: $From) -> Self {
+                <$To>::try_from(value).unwrap_or_else(|_| {
+                    panic!(
+                        "Failed to convert from {}: {:?} to {}",
```
Comment (pwang7) [inline src/fuse/conversion.rs]: try not to use {:?}, use {} instead, clippy might give warning to {:?}

File: src/memfs.rs [hunk 2]
```diff
@@ -34,7 +34,7 @@ mod util {
 
     pub fn parse_oflag(flags: u32) -> OFlag {
         debug_assert!(
-            flags < std::i32::MAX as u32,
+            flags < u32::cast(std::i32::MAX),
```
Comment (pwang7) [inline src/memfs.rs]: can you do `std::i32::MAX.cast()`?

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/memfs.rs [hunk 3]
```diff
@@ -808,7 +808,7 @@ impl INode {
         let ino = attr.ino;
         let file_data = file_node.data.get_mut();
 
-        let size_after_write = offset as usize + data.len();
+        let size_after_write = Cast::<usize>::cast(offset) + data.len();
```
Comment (pwang7) [inline src/memfs.rs]: why not `offset.cast()`?

File: src/memfs.rs [hunk 4]
```diff
@@ -1339,7 +1340,7 @@ impl Filesystem for MemoryFilesystem {
                         under the directory of ino={}",
                     child_name,
                     child_ino,
-                    offset + i as i64 + 1,
+                    offset + Cast::<i64>::cast(i) + 1,
```
Comment (pwang7) [inline src/memfs.rs]: why not `i.cast()`?

File: src/memfs.rs [hunk 5]
```diff
@@ -1325,11 +1326,11 @@ impl Filesystem for MemoryFilesystem {
 
         let readdir_helper = |data: &BTreeMap<OsString, DirEntry>| {
             let mut num_child_entries = 0;
-            for (i, (child_name, child_entry)) in data.iter().enumerate().skip(offset as usize) {
+            for (i, (child_name, child_entry)) in data.iter().enumerate().skip(offset.cast()) {
                 let child_ino = child_entry.ino;
                 reply.add(
                     child_ino,
-                    offset + i as i64 + 1, // i + 1 means the index of the next entry
+                    offset + Cast::<i64>::cast(i) + 1, // i + 1 means the index of the next entry
```
Comment (pwang7) [inline src/memfs.rs]: why not `i.cast()`?

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/memfs.rs (hunk 3 above)
Comment (pzheng1025) [reply src/memfs.rs]: compile cannot infer the type in this case.

File: src/fuse/conversion.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+use std::convert::TryFrom;
+
+/// A type cast trait used to replace as conversion.
+pub trait Cast {
```
Comment (pwang7) [inline src/fuse/conversion.rs]: pub(crate)

File: src/fuse/mod.rs [hunk 7]
```diff
@@ -32,11 +32,13 @@ pub use mount::options_validator;
 mod abi;
 mod argument;
 mod channel;
+mod conversion;
 mod ll_request;
 mod mount;
 mod reply;
 mod request;
 mod session;
+pub use conversion::Cast;
```
Comment (pwang7) [inline src/fuse/mod.rs]: is `pub use` needed?

File: src/memfs.rs [hunk 8]
```diff
@@ -1280,11 +1280,11 @@ impl Filesystem for MemoryFilesystem {
         );
 
         let read_helper = |content: &Vec<u8>| {
-            if (offset as usize) < content.len() {
-                let read_data = if ((offset + size as i64) as usize) < content.len() {
-                    &content[(offset as usize)..(offset + size as i64) as usize]
+            if offset.cast::<usize>() < content.len() {
+                let read_data = if (offset + size.cast::<i64>()).cast::<usize>() < content.len() {
```
Comment (pwang7) [inline src/memfs.rs]: offset.cast() + size < content.len()
no need extra cast()

File: src/memfs.rs [hunk 9]
```diff
@@ -1280,11 +1280,11 @@ impl Filesystem for MemoryFilesystem {
...
+                    &content[offset.cast()..(offset + size.cast::<i64>()).cast()]
```
Comment (pwang7) [inline src/memfs.rs]: offset.cast() + size
no need extra cast()

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/memfs.rs (hunk 8 above)
Comment (pzheng1025) [reply src/memfs.rs]: size is u32, content.len() is usize. So need two casts

File: src/memfs.rs (hunk 9 above)
Comment (pzheng1025) [reply src/memfs.rs]: The same

File: src/fuse/mod.rs (hunk 7 above)
Comment (pzheng1025) [reply src/fuse/mod.rs]: If you don't use pub, memfs.rs cannot use it.

File: src/memfs.rs (hunk 8 above)
Comment (pwang7) [reply src/memfs.rs]: `offset.cast::<usize>() + size.cast::<usize>() < content.len()`

File: src/memfs.rs (hunk 9 above)
Comment (pwang7) [reply src/memfs.rs]: `offset.cast::<usize>() + size.cast::<usize>()`

File: src/fuse/mod.rs (hunk 7 above)
Comment (pwang7) [reply src/fuse/mod.rs]: pub(crate) use?

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/fuse/mod.rs (hunk 7 above)
Comment (pzheng1025) [reply src/fuse/mod.rs]: Done

File: src/fuse/conversion.rs (hunk 6 above)
Comment (pzheng1025) [reply src/fuse/conversion.rs]: Done

File: src/memfs.rs (hunk 8 above)
Comment (pzheng1025) [reply src/memfs.rs]: Done

File: src/memfs.rs (hunk 9 above)
Comment (pzheng1025) [reply src/memfs.rs]: Done

**Review by pwang7 — APPROVED** [review]

**Pull request merged**